packages/flutter/lib/src/material/navigation_rail.dart

```diff
@@ -0,0 +1,849 @@
+            ),
+          ),
+          Semantics(
+            label: indexLabel,
+          ),
+        ]
+      ),
+    );
+  }
+
+  double _normalLabelFadeInValue() {
+    if (destinationAnimation.value < 0.25) {
+      return 0;
+    } else if (destinationAnimation.value < 0.75) {
+      return (destinationAnimation.value - 0.25) * 2;
+    } else {
+      return 1;
+    }
+  }
+
+  double _normalLabelFadeOutValue() {
+    if (destinationAnimation.value > 0.75) {
+      return (destinationAnimation.value - 0.75) * 4.0;
+    } else {
+      return 0;
+    }
+  }
+
+  double _extendedLabelFadeValue() {
+    return extendedTransitionAnimation.value < 0.25 ? extendedTransitionAnimation.value * 4.0 : 1.0;
+  }
+}
+
+/// Defines the behavior of the labels of a [NavigationRail].
+///
+/// See also:
+///
+///   * [NavigationRail]
+enum NavigationRailLabelType {
+  /// Only the icons of a navigation rail item are shown.
+  none,
+
+  /// Only the selected navigation rail item will show its label.
+  ///
+  /// The label will animate in and out as new items are selected.
+  selected,
+
+  /// All navigation rail items will show their label.
+  all,
+}
+
+/// A description for an interactive button within a [NavigationRail].
+///
+/// See also:
+///
+///  * [NavigationRail]
+class NavigationRailDestination {
+  /// Creates a destination that is used with [NavigationRail.destinations].
+  ///
+  /// [icon] should not be null and [label] should not be null when this
+  /// destination is used in the [NavigationRail].
+  const NavigationRailDestination({
+    @required this.icon,
+    Widget activeIcon,
+    this.label,
+  }) : activeIcon = activeIcon ?? icon,
+       assert(icon != null);
+
+  /// The icon of the destination.
+  ///
+  /// Typically the icon is an [Icon] or an [ImageIcon] widget. If another type
+  /// of widget is provided then it should configure itself to match the current
+  /// [IconTheme] size and color.
+  ///
+  /// If [activeIcon] is provided, this will only be displayed when the
+  /// destination is not selected.
+  ///
+  /// To make the [NavigationRail] more accessible, consider choosing an
+  /// icon with a stroked and filled version, such as [Icons.cloud] and
+  /// [Icons.cloud_queue]. [icon] should be set to the stroked version and
+  /// [activeIcon] to the filled version.
+  final Widget icon;
+
+  /// An alternative icon displayed when this destination is selected.
+  ///
+  /// If this icon is not provided, the [NavigationRail] will display [icon] in
+  /// either state.
+  ///
+  /// See also:
+  ///
+  ///  * [NavigationRailDestination.icon], for a description of how to pair
+  ///    icons.
+  final Widget activeIcon;
+
+  /// The label for the destination.
+  ///
+  /// The label should be provided when used with the [NavigationRail]. When
+  /// the labelType is [NavigationRailLabelType.none] and the rail is not
+  /// extended, then it can be null, but should be used for semantics.
```

Comment:
Not really clear what "then it can be null, but should be used for semantics" means

Reply:
Updated